src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -337,87 +354,279 @@ out ImmutableArray<(BoundExpression, BoundDagTemp)> bindings)
+                    if (d != dd && !killsDecisionOnTrueBranch) trueBuilder?.Add(dd);
+                }
+                else
+                {
+                    trueBuilder?.Free();
+                    trueBuilder = null;
+                }
+                if (permitsFalse)
+                {
+                    Debug.Assert(d != dd);
+                    if (!killsDecisionOnFalseBranch) falseBuilder?.Add(dd);
+                }
+                else
                 {
-                    case BoundDagFieldEvaluation e: return e.Field;
-                    case BoundDagPropertyEvaluation e: return e.Property;
-                    case BoundDagTypeEvaluation e: return e.Type;
-                    case BoundDagDeconstructEvaluation e: return e.DeconstructMethod;
-                    default: throw ExceptionUtilities.UnexpectedValue(this.Kind);
+                    falseBuilder?.Free();
+                    falseBuilder = null;
                 }
             }
+
+            if (trueBuilder != null)
+            {
+                var pcd = trueBuilder.Count == c.Decisions.Length ? c : new PartialCaseDecision(c.Index, c.Syntax, trueBuilder.ToImmutableAndFree(), c.Bindings, c.WhereClause, c.CaseLabel);
+                whenTrueBuilder.Add(pcd);
+            }
+
+            if (falseBuilder != null)
+            {
+                var pcd = falseBuilder.Count == c.Decisions.Length ? c : new PartialCaseDecision(c.Index, c.Syntax, falseBuilder.ToImmutableAndFree(), c.Bindings, c.WhereClause, c.CaseLabel);
+                whenFalseBuilder.Add(pcd);
+            }
         }
-        public override int GetHashCode()
-        {
-            return this.Input.GetHashCode() ^ (this.Symbol?.GetHashCode() ?? 0);
-        }
-        public static bool operator ==(BoundDagEvaluation left, BoundDagEvaluation right)
-        {
-            return (left == (object)null) ? right == (object)null : left.Equals(right);
-        }
-        public static bool operator !=(BoundDagEvaluation left, BoundDagEvaluation right)
-        {
-            return !left.Equals(right);
-        }
-    }
 
-    partial class BoundDagTemp
-    {
-        public override bool Equals(object obj) => obj is BoundDagTemp other && this.Equals(other);
-        public bool Equals(BoundDagTemp other)
+        /// <summary>
+        /// Given that the decision d has occurred and produced a true/false result,
+        /// set permitsTrue if a true decision on d would permit other to succeed.
+        /// set permitsFalse if a false decision on d would permit other to succeed.
+        /// sets killsFalseDecision when d being false means other has been proven true
+        /// </summary>
+        private void CheckConsistentDecision(
+            BoundDagDecision d,
+            BoundDagDecision other,
+            out bool permitsTrue,
+            out bool permitsFalse,
+            out bool killsDecisionOnTrueBranch,
+            out bool killsDecisionOnFalseBranch)
```

Comment:
>killsDecisionOnFalseBranch [](start = 21, length = 26)

It feels like the logic in this method and its consumer could be much easier to follow if it returned a single value - a combination of bits:
couldBeTrueIfTrue
couldBeFalseIfTrue
couldBeTrueIfFalse
couldBeFalseIfFalse

We start with all four bits set and simply erase those that proven to be false.

Reply:
Two of them have the opposite sense... isDefinitelyTrueIfTrue, and isDefinitelyFalseIfFalse. I'm going to keep it this way for now, but take a note (in the code) that it could be reorganized and/or renamed for clarity.

Reply:
> Two of them have the opposite sense... isDefinitelyTrueIfTrue, and isDefinitelyFalseIfFalse.

Yes, and this is what makes it hard to follow, at least for me. "isDefinitelyTrueIfTrue" is couldBeTrueIfTrue is set and couldBeFalseIfTrue is not set

Reply:
Noted/addressed in https://github.com/dotnet/roslyn/pull/24059/commits/114a56e9077ed03f08443889b69cae6972a4acc6 of https://github.com/dotnet/roslyn/pull/24059